pilot/pkg/proxy/envoy/v2/mesh_discovery.go

```diff
@@ -0,0 +1,219 @@
+	}
+}
+
+// IsEnabled returns true if EnvoyV2 APIs are enabled for Pilot.
+func IsEnabled() bool {
+	return len(gRPCPort) > 0
+}
+
+// NewDiscoveryServer creates DiscoveryServer that sources data from Pilot's internal mesh data structures
+func NewDiscoveryServer(mesh MeshDiscovery) *DiscoveryServer {
+	// TODO for now use hard coded / default gRPC options. The constructor may evolve to use interfaces that guide specific options later.
+	// Example:
+	//		grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(uint32(someconfig.MaxConcurrentStreams)))
+	var grpcOptions []grpc.ServerOption
+
+	var interceptors []grpc.UnaryServerInterceptor
+
+	// TODO: log request interceptor if debug enabled.
+
+	// setup server prometheus monitoring (as final interceptor in chain)
+	interceptors = append(interceptors, grpc_prometheus.UnaryServerInterceptor)
+	grpc_prometheus.EnableHandlingTimeHistogram()
+
+	grpcOptions = append(grpcOptions, grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(interceptors...)))
+
+	// get the grpc server wired up
+	grpc.EnableTracing = true
+
+	grpcServer := grpc.NewServer(grpcOptions...)
+	out := &DiscoveryServer{mesh: mesh, grpcServer: grpcServer}
+	xdsapi.RegisterEndpointDiscoveryServiceServer(out.grpcServer, out)
+	return out
+}
+
+// Start opens up a listening port for serving gRPC requests and starts up the underlying gRPC server.
+// This method is non-blocking.
+func (s *DiscoveryServer) Start() {
+	lis, err := net.Listen("tcp", ":"+gRPCPort)
+	if err != nil {
+		log.Errorf("mesh discovery server failed to listen on port %q: %v", gRPCPort, err)
+		return
+	}
+	// Serve is a blocking call.
+	go func(gRPCServer *grpc.Server, lis net.Listener) {
+		log.Infof("attempting to start mesh discovery server on port %q", gRPCPort)
+		if err := gRPCServer.Serve(lis); err != nil {
+			log.Errorf("mesh discovery server failed to serve on port %q: %v", gRPCPort, err)
+		}
+	}(s.grpcServer, lis)
+}
+
+/***************************  Mesh EDS Implementation **********************************/
+
+// StreamEndpoints implements xdsapi.EndpointDiscoveryServiceServer.StreamEndpoints().
+func (s *DiscoveryServer) StreamEndpoints(stream xdsapi.EndpointDiscoveryService_StreamEndpointsServer) error {
+	ticker := time.NewTicker(responseTickDuration)
+	peerInfo, ok := peer.FromContext(stream.Context())
+	peerAddr := "Unknown peer address"
+	if ok {
+		peerAddr = peerInfo.Addr.String()
+	}
+	defer ticker.Stop()
+	var discReq xdsapi.DiscoveryRequest
+	var receiveError error
+	var clusters []string
+	initialRequest := true
+	reqChannel := make(chan xdsapi.DiscoveryRequest, 1)
+	go func() {
+		defer close(reqChannel)
+		for {
+			req, err := stream.Recv()
+			if err != nil {
+				if status.Code(err) == codes.Canceled || err == io.EOF {
+					return
+				}
+				receiveError = err
+				log.Errorf("request loop for EDS for client %q terminated with errors %v", peerAddr, err)
+				return
+			}
+			reqChannel <- *req
+		}
+	}()
+	for {
```

Comment:
Chatted with Gautam: I think this can go in as-is. Down the line when we land other xDS endpoints, I think we'll want to extract out a common pattern:
1. Read the initial request from the stream
2. Use the initial request's data to init serving to this Envoy instance
3. Start concurrent loops to drain requests and send responses.

This general pattern should be common across the xDS endpoints, so it'd be nice to reify it if possible once we have more than the single implementation.